dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,190 @@
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotMetrics:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        project_id: The ID of the Project
+        n_unique_users: The number of unique users who downloaded from the Project
+        last_download_date: The date of the last download from the Project
+        total_data_size_in_gib: The current total size of the Project in GiB
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                ),
+                RECENT_DOWNLOADS AS (
+                    SELECT PROJECT_ID, RECORD_DATE, USER_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE DATE_TRUNC('MONTH', RECORD_DATE) = DATE_TRUNC('MONTH', DATE({context["params"]["month_to_run"]}))
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                TOP_10_PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID, COUNT(DISTINCT USER_ID) AS N_UNIQUE_USERS
+                    FROM RECENT_DOWNLOADS
+                    GROUP BY PROJECT_ID
+                    ORDER BY N_UNIQUE_USERS DESC
+                    LIMIT 10
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM TOP_10_PUBLIC_PROJECTS)
```

Comment:
We should probably add in a `is_public` check here as well because there can be private files within a public project that we want to filter activity out for.

Reply:
I decided against this because calculating a project size based off only the public files wouldn't truly reflect the size of the project itself. It would also mislead users into thinking that they have access to all of a project, when in reality they might only have access to a portion of it, if that makes sense. So right now I'm still leaning against filtering for public files. What do you think?